### ISSUE TITLE
"last run" Super NT video and audio issue with MSU-1 games

### ISSUE THREAD

**Issue by speedyrules** [analoguejb/Analogue-Super-Nt-JB#4]
Dear analoguejb,

This is a double post:
- sent via e-mail to support_at_analogue.co
- posted on here on github

I think i found an issues with the "last run" Super NT, which i guess it is related to the new hardware used in it or consequently with the firmware.
Basically the issue are MSU-1 games, which don't play well. Not all, but some. I will use "Super Road Blaster" (further mentioned as SRB) for this issue.
I and others experience video and audio issues with SRB.

My personal setup: 
- "last run" Super NT with latest JB V7.2 (further mentioned as SNT)
- revD FxPak Pro (further mentioned as FX)

Clarification:
- older SNT means: Any other version of the SNT but the "last run"
- previosRev FX means: Any other revision of the FX but revD

I will attach 2 videos to demonstrate the issue.

I know there are many possible sources of errors "in the chain". Or that i have an "individual" faulty system. 
But i think i have narrowed it down to the "last run" SNT.

Here the tests and results for SRB i went through:
- revD FX - original SNES PAL: OK
- revD FX - original SNES US: OK
- revD FX - original SFC JP: OK
- revD FX - last run SNT: NOK <- !!!
- revD FX - older SNT: OK
- previosRev FX - last run SNT: NOK <- !!!
- previosRev FX - older SNT: OK

Could you fix this please? :)

https://github.com/analoguejb/Analogue-Super-Nt-JB/assets/25831136/940c30f7-a01f-44d2-bc36-00893af0e331

https://github.com/analoguejb/Analogue-Super-Nt-JB/assets/25831136/c6041832-b513-457b-bf69-7479bc2372dd

**Comment by DCSpecial** [comment]
Can agree that these problems happen
RevB FxPak Pro on SuperNT last run same issues as described
Same Cart/Rom on 1 Chip SNES -> working fine

**Comment by Real-Dampfwalze** [comment]
I faced a similar problem (at least I had similar sounding dropouts) with my final run SuperNT but with Super Gameboy games and I found out that the final run SuperNT doesn't accept the same volume input level of external cartridge sound (which is used by MSU1 and Super Gameboy Games with the FXPak) as the old SuperNT did.

Suggestion: check if you activated the "boost MSU1 volume" in the FXPak settings. If it is not activated and still cuts the sound off, try lowering the cartridge sound volume level in the Super NT settings.

For me it fixed the Super GameBoy sound problem. Maybe your problem is  different but I guess you could give it a try. :)

**Comment by speedyrules** [comment]
Hey @Real-Dampfwalze 
Thanks for the reply. Seams that not a lot of people really care about "last run" issues.. however..

My MSU-1 Volume boost is off. Always was.
Lowering the cartridge volume level did not help.
Interestingly Star Fox plays audio almost everywhere very badly.
(works like a charm on mister.. and works well on og snes with the fxpak)

But as i learned, this is audio. From a chat with ikari, audio and video are most probably different issues.

@analoguejb any help from your side?

**Comment by henriquewicz** [comment]
Same problems here. The cart and games working fine in a original snes. but in super nt is terrible. not only with road blaster or other msu games but a lot of games dont run well too..

**Comment by MichaelW1980** [comment]
I also have a "last run" SNT and a Rev. D FX Pak Pro and a Rev X. SD2SNES (the chinese knockoff), a friend did lend to me.
I had a very similar issue. The culprit seems to be the volume of the PCM Files. Many MSU-1 assets come in too loud fo the SNT.  And the revision of the cartridge doesn't seem to make a difference.

A sound of louder than a certain level can cause the SNT to completely mute out the cartridge audio for a moment. For me, the solution was, to normalize the MSU-1 PCM files themselves to -21dBFS. Depending on the game, I have also gone as low as -24 dBFS. MSU-1 FMV has the audio part seperate, as a PCM file as well, if I'm not mistaken, so this should also work for video material just the same:

I found [this Article](https://www.zeldix.net/t2068-msu-1-re-normalization-lower-pcm-volume) on Zeldix, that shares a rather easy to use tool for Windows, to re-normalize PCM audio. I [checked the download](https://www.virustotal.com/gui/file/24d06035b0ef58e21216f9a7ce736e271ea3c5997b921e89e61a4b69291e020a) on VirusTotal just to be sure and it appears to be clean.

Simply download the whole zip file, extract it to some folder of your choice, put the PCM Files to the In subdirectory and run the software. Type in -21 (or -24) and hit enter. The tool will do the normalizing alll by itself. After the tool is done, use the files from the out folder and give it another go on the FX.

**Comment by Real-Dampfwalze** [comment]
Thank you for this work around. 
I still think it is a bug/problem of the SuperNT, because the older revision and a original SNES accept this loudness fine.

Question is: is it possilbe to fix by a firmware update, or is it a flaw in the hardware design.

If a software update could fix it, it would be great of Analogue to fix that problem instead of me having to build extra versions for the rev. 2 SuperNT. 😊

**Comment by speedyrules** [comment]
> The culprit seems to be the volume of the PCM Files.

This might be one part of the problem. Ikari suggested to try this, what i did. Dis not solve the problem though.
Happy it works for you though!

On top there is also a video issue as you can see in my uploaded video.
So it might be also combination of aoftware and hardware issues.

Hope dies last.. but not sure this will ever be fixed.

**Comment by mrehkopf** [comment]
In April of 2024 I had an opportunity to try this on a last-run Super Nt myself and capture a logic trace. I also analyzed the issue of audio cutting out.
I'm not sure the stuttering/flickering issue with Super Road Blaster has been fixed in the meantime, anyway here are my findings.

### Audio cutting out
This is caused by pre-amplification of the cartridge line signal being too hot for the new audio codec IC used as the ADC. The amplitude of the ADC input signal becomes too large so it even drives up the IC's internal reference voltage, drowning the ADC. The IC has analog audio *output* pins as well, which are not used for purposes of the Super Nt as far as I can tell but serve to illustrate the effect. Here's an oscilloscope shot of the issue occurring:

<img width="1280" height="824" alt="Image" src="https://github.com/user-attachments/assets/8b80921f-c65f-435f-976e-fcf86c0cbbd2" />

**Yellow**: the signal coming in at the cartridge slot (one of the two audio channels from the cartridge) at roughly 1.6Vp-p. It goes to a TL9062 OPAMP where it is amplified.
**Orange**: the pre-amplified signal going into the audio codec IC/ADC (Realtek ALC5616). Gain is >2, it looks like the boosted signal is at about ~3.6Vp-p.
**Periwinkle (sorry)**: the analog output signal of the ALC5616. It's the same signal that goes through ADC processing internally.
**Green**: the internal reference voltage for the ADC. It stays constant at first (at ~0.9V which is correctly half the recommended power supply voltage of 1.8V) but as soon as the audio input amplitude rises it gets "pumped" up as well, almost up to 2V, and when it crosses a certain threshold the ADC audio signal (peri-blue-ish-lavender) starts drowning until it goes completely flat.

This is purely an analog issue which can't be fixed in firmware, the only remedy is lowering the pre-amplification gain at the TL9062, probably by altering the values of R117, R119, R123 / R120, R121, R122, or keeping the cartridge audio signal below a certain level.

### Video stuttering/flickering
This is a behavioral issue of the SNES implementation which should be possible to fix in firmware but I don't know why it only occurs with the last-run model; it's an FPGA implementation detail and all models should use the same implementation... oh well. Maybe not.

The issue is: 
Super Road Blaster disables V-Blank NMI near the bottom of the screen, and sometimes re-enables it only when already inside the VBlank period. ($4200.7 = 1). The expected SNES behaviour is that the NMI is then triggered immediately. **The last-run Super Nt does not do that**. When NMI is enabled after the non-VBlank-to-VBlank transition, it simply does not fire. That throws off the video player timing, it does not buffer/refresh often enough and therefore stutters, and image flickering is triggered by other code having enabled color windowing which should be disabled again by the NMI code.

tl;dr: Last-Run Super Nt does not fire NMI when it is re-enabled during VBlank, but it should.
(I can't imagine that doesn't affect some other games as well.)

Relevant logic traces:

| <img width="1320" height="273" alt="Image" src="https://github.com/user-attachments/assets/3d140148-76f9-41a4-8567-ddae263c4cd4" /> |
|:--:|
| Logic trace of intact NMI triggering (cursor 2) after $4200 write (cursor 1). Note the NMI vector is patched to $2A10 due to the sd2snes's in-game hook being enabled. The original vector would be $8000. |

| <img width="1320" height="273" alt="Image" src="https://github.com/user-attachments/assets/346dcc0f-d31a-4d3f-88e4-62dd466e2ea0" /> |
|:--:|
| Logic trace from the last-run Super Nt of the same MSU video frame; Cursor 1 shows the $4200 write but NMI entry is absent; code continues to execute inside the IRQ handler. |
